File: tools/editor/app/assets.js
/* Client half of the asset path: the dev server hands over a sprite manifest
 * extracted live from BattleshipsForever.exe, and this fetches frames on demand.
 *
 * Two things here are not optional, both of them traps the design note calls out:
 *
 *  1. Half the game's art is a PALETTE MASK — three shades of grey the engine
 *     multiplies by a colour at draw time. spr_Nebula, spr_Spacestation,
 *     spr_Meteor, spr_Marker and every hull section are masks. Blit one straight
 *     and you get a black box, so masks go through tinted().
 *  2. ORIGINS ARE DATA. spr_MesHQ's origin is (0,0), spr_Rock's is its centre.
 *     Everything is drawn at -ox,-oy, never at -w/2,-h/2.
 *
 * Tints below are the ones the objects set in their own Create events, read out
 * of the decrypted tree — not taste.
 */
window.Assets = (function () {
  'use strict';

  // bfdefault.ini [<team> Colours]; index 0 is the hull default.
  const TEAMS = [[[0, 255, 0], [64, 255, 64], [128, 255, 128]],
  [[0, 255, 255], [64, 128, 128], [128, 196, 196]],
  [[255, 0, 0], [255, 64, 64], [155, 45, 45]],
  [[255, 0, 255], [128, 0, 255], [255, 128, 128]]];

  const TINT = {                     // object -> image_blend / l_colour at create
    MoveToArea: 'rgb(0,255,0)', Marker: 'rgb(0,255,0)',      // image_blend = c_green
    SpaceStation: 'rgb(0,255,255)',                          // l_colour = global.colour[1,0]
    obs_Meteor: 'rgb(255,255,255)',                          // l_colour = c_white
    ter_Nebula: 'rgb(160,190,255)', ter_MegaNebula: 'rgb(160,190,255)',
    ter_Sun: 'rgb(255,235,180)',
  };
  const ALPHA = { ter_Nebula: 0.38, ter_MegaNebula: 0.5, ter_Sun: 0.9 };

  const A = {
    ok: false, sprites: {}, objects: {}, source: null,
    hulls: new Set(),                // object names that are ships, not props
    onload: null,                    // set by the app: called when art arrives
    _img: {}, _tint: new Map(), _ship: {},
  };

  A.init = function () {
    return fetch('api/assets').then((r) => r.json()).then((d) => {
      if (!d.ok) throw new Error(d.why || 'no assets');
      A.ok = true; A.sprites = d.sprites; A.objects = d.objects; A.source = d.source;
      A.hulls = new Set(d.hulls || []);
      return A;
    }).catch((e) => { A.ok = false; A.why = e.message; return A; });
  };

  /* Ship or prop, answered now.
   *
   * The distinction the UI turns on, and the one the sprite manifest cannot
   * make: a ship draws its sections, so its `sprite_index` is spr_Core and
   * every hull in the game shares it. `A.ship()` knows the answer too, but only
   * after a round trip, and a picker listing 600 objects has to decide how to
   * draw each of them before it has fetched any. So the server sends the whole
   * set with the manifest and this is a lookup. */
  A.isShip = (obj) => A.hulls.has(obj);

  /* Lazily fetched frame. Returns null until it has arrived, then repaints. */
  A.img = function (name, frame) {
    frame = frame || 0;
    const s = A.sprites[name];
    if (!s) return null;
    const key = name + '#' + (frame % s.n);
    let im = A._img[key];
    if (im === undefined) {
      im = new Image();
      im.onload = () => { im._ready = true; if (A.onload) A.onload(); };
      im.onerror = () => { A._img[key] = null; };
      im.src = 'api/sprite/' + encodeURIComponent(name) + '?f=' + (frame % s.n);
      A._img[key] = im;
    }
    return im && im._ready ? im : null;
  };

  /* mask × colour, exactly as image_blend does — the one place the composite is
     written. Fill with the colour, `multiply` the art in, then `destination-in`
     to recover the art's own alpha. Three call sites needed this (two here, one
     in ships.js) and three transcriptions of a three-op sequence is three
     chances for one of them to disagree about the order.

     Takes a ready image or canvas and returns a fresh canvas; caching is the
     caller's, because each of them keys it differently. */
  A.maskTint = function (src, w, h, rgb) {
    const c = document.createElement('canvas');
    c.width = w; c.height = h;
    const g = c.getContext('2d');
    g.fillStyle = rgb; g.fillRect(0, 0, w, h);
    g.globalCompositeOperation = 'multiply';
    g.drawImage(src, 0, 0);
    g.globalCompositeOperation = 'destination-in';
    g.drawImage(src, 0, 0);
    return c;
  };

  /* Cached per (sprite, frame, rgb). */
  function tinted(name, frame, rgb) {
    const key = name + '#' + frame + '|' + rgb;
    let c = A._tint.get(key);
    if (c) return c;
    const im = A.img(name, frame), s = A.sprites[name];
    if (!im) return null;
    c = A.maskTint(im, s.w, s.h, rgb);
    A._tint.set(key, c);
    return c;
  }
  A.tinted = tinted;

  A.spriteOf = (obj) => A.objects[obj] || null;

  /* Draw one sprite frame with its real origin, in world units.
     opts: {frame, angle (GM degrees, ccw), scale, tint, alpha} */
  A.blit = function (ctx, name, x, y, opts) {
    opts = opts || {};
    const s = A.sprites[name];
    if (!s) return false;
    const f = opts.frame || 0;
    const src = opts.tint ? tinted(name, f % s.n, opts.tint) : A.img(name, f);
    if (!src) return false;
    const k = opts.scale == null ? 1 : opts.scale;
    ctx.save();
    ctx.globalAlpha = opts.alpha == null ? 1 : opts.alpha;
    ctx.translate(x, y);
    if (opts.angle) ctx.rotate(-opts.angle * Math.PI / 180);   // GM angles are ccw
    ctx.scale(k, k);
    ctx.drawImage(src, -s.ox, -s.oy);
    ctx.restore();
    return true;
  };

  /* Draw an object the way the game would: its sprite, its blend, its alpha. */
  A.drawObject = function (ctx, obj, x, y, opts) {
    opts = opts || {};
    const name = A.spriteOf(obj);
    if (!name || !A.sprites[name]) return false;
    // the blend the object sets at create wins; a mask with no declared blend
    // still has to go through white, which is what image_blend defaults to
    return A.blit(ctx, name, x, y, Object.assign({
      tint: TINT[obj] || (A.sprites[name].mask ? 'rgb(255,255,255)' : null),
      alpha: ALPHA[obj],
    }, opts));
  };

  /* Hulls are not in the exe in readable form; the server serves the render-model
     dump when it has one. Returns null until loaded (and never, if there is none). */
  A.ship = function (name) {
    if (name in A._ship) return A._ship[name];
    A._ship[name] = null;
    fetch('api/ship/' + encodeURIComponent(name))
      .then((r) => (r.ok ? r.json() : null))
      .then((d) => {
        if (!d) return;
        d._img = {};
        Object.entries(d.sprites).forEach(([nm, sp]) => {
          const im = new Image();
          im.onload = () => { im._ready = true; if (A.onload) A.onload(); };
          im.src = sp.src;
          d._img[nm] = im;
        });
        A._ship[name] = d;
        if (A.onload) A.onload();
      })
      .catch(() => { });
    return null;
  };

  /* A real hull: every section blitted at its own offset, angle and team shade. */
  A.drawShip = function (ctx, hull, x, y, opts) {
    opts = opts || {};
    const team = opts.team == null ? 0 : opts.team;
    const angle = opts.angle || 0;
    ctx.save();
    ctx.translate(x, y);
    ctx.rotate(-angle * Math.PI / 180);
    let drew = 0;
    (hull.sections || []).forEach((sec) => {
      const sp = hull.sprites[sec.sprite], im = hull._img[sec.sprite];
      if (!sp || !im || !im._ready) return;
      const rgb = TEAMS[team][sec.cls == null ? 0 : sec.cls];
      const key = sec.sprite + '|' + rgb + '|' + team;
      let src = A._tint.get(key);
      if (!src) {
        // Unmasked art is already what it should look like — cache the image
        // itself rather than a canvas holding a copy of it.
        src = sp.mask
          ? A.maskTint(im, sp.w * sp.n, sp.h, `rgb(${rgb[0]},${rgb[1]},${rgb[2]})`)
          : im;
        A._tint.set(key, src);
      }
      ctx.save();
      ctx.translate(sec.x, sec.y);
      ctx.rotate(-(sec.angle || 0) * Math.PI / 180);
      ctx.scale(sec.xscale == null ? 1 : sec.xscale, sec.yscale == null ? 1 : sec.yscale);
      ctx.drawImage(src, 0, 0, sp.w, sp.h, -sp.ox, -sp.oy, sp.w, sp.h);
      ctx.restore();
      drew++;
    });
    ctx.restore();
    return drew > 0;
  };

  /* Extent of a hull, in its own units.

     A design file has a bbox because scene.py computes one; a stock hull dump
     does not, so it is measured here — each section's sprite rect carried
     through the same translate/rotate/scale drawShip applies to it, which is
     why the order below is scale, then rotate, then translate. Cached on the
     hull object, since that is what the fetch already hands back. */
  function hullBox(hull) {
    if (hull._box !== undefined) return hull._box;
    let x0 = Infinity, y0 = Infinity, x1 = -Infinity, y1 = -Infinity;
    (hull.sections || []).forEach((sec) => {
      const sp = hull.sprites[sec.sprite];
      if (!sp) return;
      const a = -(sec.angle || 0) * Math.PI / 180;
      const co = Math.cos(a), si = Math.sin(a);
      const kx = sec.xscale == null ? 1 : sec.xscale;
      const ky = sec.yscale == null ? 1 : sec.yscale;
      [[-sp.ox, -sp.oy], [sp.w - sp.ox, -sp.oy],
        [-sp.ox, sp.h - sp.oy], [sp.w - sp.ox, sp.h - sp.oy]].forEach((c) => {
        const px = c[0] * kx, py = c[1] * ky;
        const x = sec.x + px * co - py * si, y = sec.y + px * si + py * co;
        if (x < x0) x0 = x;
        if (x > x1) x1 = x;
        if (y < y0) y0 = y;
        if (y > y1) y1 = y;
      });
    });
    hull._box = x1 > x0 ? [x0, y0, x1, y1] : null;
    return hull._box;
  }

  /* A hull thumbnail for a ship *object* — the counterpart of Ships.thumb, which
     does the same job for a design file. Returns false until the dump and its
     art have arrived; every caller repaints on A.onload, and a repaint that
     happens twice is free. */
  A.shipThumb = function (canvas, obj, pad, team) {
    const ctx = canvas.getContext('2d');
    ctx.clearRect(0, 0, canvas.width, canvas.height);
    const hull = A.ship(obj);
    if (!hull) return false;
    const box = hullBox(hull);
    if (!box) return false;
    const w = Math.max(1, box[2] - box[0]), h = Math.max(1, box[3] - box[1]);
    const m = pad == null ? 4 : pad;
    const k = Math.min((canvas.width - m * 2) / w, (canvas.height - m * 2) / h);
    ctx.save();
    ctx.translate(canvas.width / 2 - ((box[0] + box[2]) / 2) * k,
      canvas.height / 2 - ((box[1] + box[3]) / 2) * k);
    ctx.scale(k, k);
    const drew = A.drawShip(ctx, hull, 0, 0, { team: team || 0 });
    ctx.restore();
    return drew;
  };

  A.TEAMS = TEAMS;
  return A;
})();

File: tools/editor/assets.py
#!/usr/bin/env python3
"""Sprites out of BattleshipsForever.exe, in memory, on demand.

Nothing is ever written to disk: the exe is decrypted into RAM at startup
(~50 ms for the whole 6.9 MB tree), sprite headers are indexed, and a frame is
only inflated and PNG-encoded the first time something asks for it. That keeps
derived game art out of the repo by construction — there is no file to commit.

GM 7.0 sprite record, verified against every sprite in the v0.90d tree:

    [u32 len][name][u32 542]
    [i32 w][i32 h]
    [i32 bbox_l][i32 bbox_r][i32 bbox_b][i32 bbox_t]
    [u32 transparent][u32 smooth][u32 preload][u32 bbox_mode][u32 precise]
    [i32 origin_x][i32 origin_y]            <-- data, never assume centred
    [u32 n_frames]
    n × ( [u32 540][u32 1][u32 w][u32 h][u32 len][zlib raw BGRA] )

Origins really do vary: spr_Rock is (24,24) on a 48×48, spr_MesHQ is (0,0).

Usage as a script (debug):  python3 assets.py [game-dir] [sprite-name]
"""
import os
import re
import struct
import sys
import zlib

HERE = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, os.path.dirname(HERE))          # tools/, for gm7.py

SPRITE_VER = 542
FRAME_VER = 540
NAME_RE = re.compile(rb'^[A-Za-z_][A-Za-z0-9_]{2,40}$')

# The install is discovered, never hardcoded: an absolute path embeds whoever's
# account it was authored on. tools/bsf/paths.py is the module that answers this
# for the whole repo — $BSF_GAME, then $BSF_BASE, then a search near the
# checkout, validated by ShipMaker.exe — and it documents that nothing raises at
# import time, which is what makes it safe to read at module scope here.
#
# Not campaign/build.py, which resolves the same two directories: that is the
# YAML mission compiler, it needs PyYAML, and it sys.exit()s at import when
# PyYAML is absent. A sprite reader that decrypts the exe has no business
# failing over a missing YAML parser.
sys.path.insert(0, os.path.join(os.path.dirname(HERE), 'bsf'))
import paths                                                        # noqa: E402

INSTALL = str(paths.GAME)
RESEARCH = str(paths.GAME.parent)


def u32(b, p):
    return struct.unpack_from('<I', b, p)[0]


def i32(b, p):
    return struct.unpack_from('<i', b, p)[0]


def png_encode(w, h, bgra):
    """BGRA rows -> PNG bytes. Channel swap by stride slicing, not per pixel."""
    a = bytearray(bgra)
    a[0::4], a[2::4] = a[2::4], a[0::4]
    stride = w * 4
    raw = bytearray()
    for y in range(h):
        raw.append(0)                       # filter: none
        raw += a[y * stride:(y + 1) * stride]

    def chunk(tag, data):
        return (struct.pack('>I', len(data)) + tag + data
                + struct.pack('>I', zlib.crc32(tag + data) & 0xffffffff))

    return (b'\x89PNG\r\n\x1a\n'
            + chunk(b'IHDR', struct.pack('>IIBBBBB', w, h, 8, 6, 0, 0, 0))
            + chunk(b'IDAT', zlib.compress(bytes(raw), 6))
            + chunk(b'IEND', b''))


def is_palette_mask(bgra):
    """Section art is a 3-shade grey mask the game multiplies by the team colour.
    Drawn naively it is a black box, so the client has to know which is which.
    Strided slices, not a per-pixel loop — this runs over every sprite at start."""
    return bgra[0::4] == bgra[1::4] == bgra[2::4]


class Library:
    """Sprite index over a decrypted tree. Headers eagerly, pixels lazily."""

    def __init__(self, tree):
        self.tree = tree
        self.sprites = {}
        self._png = {}
        self._scan()

    @classmethod
    def from_game(cls, game_dir=None):
        import gm7
        d = os.path.expanduser(game_dir) if game_dir else INSTALL
        exe = os.path.join(d, 'BattleshipsForever.exe')
        if not os.path.exists(exe):
            raise FileNotFoundError(
                f'BattleshipsForever.exe not found at {exe} — point $BSF_GAME '
                'at a v0.90d install, or put one beside this checkout')
        _raw, (_pos, _clen, blob) = gm7.load(exe)
        plain, *_ = gm7.gmkrypt_decrypt(blob)
        lib = cls(plain)
        lib.source = exe
        return lib

    def _scan(self):
        b = self.tree
        marker = struct.pack('<I', SPRITE_VER)
        pos = 0
        while True:
            hit = b.find(marker, pos)
            if hit < 0:
                break
            pos = hit + 4
            for n in range(3, 42):                       # walk back over the name
                start = hit - n - 4
                if start < 0:
                    break
                if u32(b, start) != n:
                    continue
                name = b[start + 4:start + 4 + n]
                if not NAME_RE.match(name):
                    break
                try:
                    rec = self._parse(hit + 4, name.decode())
                except (struct.error, ValueError):
                    rec = None
                if rec:
                    self.sprites[rec['name']] = rec
                break

    def _parse(self, p, name):
        b = self.tree
        w, h = i32(b, p), i32(b, p + 4)
        if not (0 < w <= 4096 and 0 < h <= 4096):
            return None
        p += 8 + 16 + 20                                  # bbox (4) + flags (5)
        ox, oy = i32(b, p), i32(b, p + 4)
        p += 8
        nframes = u32(b, p)
        p += 4
        if not (0 < nframes <= 512):
            return None
        frames = []
        for _ in range(nframes):
            if u32(b, p) != FRAME_VER:
                return None
            p += 8                                        # frame version + flag
            fw, fh, ln = u32(b, p), u32(b, p + 4), u32(b, p + 8)
            p += 12
            if fw != w or fh != h or not (0 < ln <= len(b) - p):
                return None
            frames.append((p, ln))
            p += ln
        return {'name': name, 'w': w, 'h': h, 'ox': ox, 'oy': oy,
                'n': nframes, 'frames': frames, 'mask': None}

    # ---------------------------------------------------------------- output
    def bgra(self, name, frame=0):
        s = self.sprites[name]
        off, ln = s['frames'][frame % s['n']]
        d = zlib.decompress(self.tree[off:off + ln])
        if len(d) != s['w'] * s['h'] * 4:
            raise ValueError(f'{name}: frame {frame} is not {s["w"]}x{s["h"]} BGRA')
        return d

    def png(self, name, frame=0):
        key = (name, frame)
        if key not in self._png:
            s = self.sprites[name]
            d = self.bgra(name, frame)
            if s['mask'] is None:
                s['mask'] = is_palette_mask(d)
            self._png[key] = png_encode(s['w'], s['h'], d)
        return self._png[key]

    def classify(self):
        """Decide mask-vs-colour for every sprite (frame 0 only). Needed before
        the client draws anything, so it is worth the one-off inflate."""
        for nm, s in self.sprites.items():
            if s['mask'] is None:
                try:
                    s['mask'] = is_palette_mask(self.bgra(nm, 0))
                except Exception:                                    # noqa: BLE001
                    s['mask'] = False
        return self

    def manifest(self):
        return {nm: {'w': s['w'], 'h': s['h'], 'ox': s['ox'], 'oy': s['oy'],
                     'n': s['n'], 'mask': s['mask']}
                for nm, s in self.sprites.items()}


# --------------------------------------------------------------------------
# object -> sprite name. The object record carries a sprite *index*, and index
# order is not file order (see tools/gmobj.py), so resolving it from the exe
# alone means walking both sections properly. The runtime dump the game itself
# writes already has the answer, so the prototype reads that when it is there
# and falls back to the handful EP9 needs.
# --------------------------------------------------------------------------
RT_DUMP = os.path.join(INSTALL, 'mods', 'objects_rt.txt')

FALLBACK_OBJ_SPRITE = {
    'Hestia': 'spr_Core', 'SpaceStation': 'spr_Spacestation', 'ter_Planet': 'spr_Planet',
    'ter_Nebula': 'spr_Nebula', 'ter_MegaNebula': 'spr_Nebula', 'ter_Rock': 'spr_Rock',
    'ter_Sun': 'spr_Sun', 'obs_Meteor': 'spr_Meteor', 'MoveToArea': 'spr_Marker',
    'Marker': 'spr_Marker', 'Civilian': 'spr_Civilian', 'EPirate06': 'spr_Pirate',
}


def object_sprites(path=RT_DUMP):
    out = dict(FALLBACK_OBJ_SPRITE)
    if os.path.exists(path):
        for ln in open(path, encoding='latin-1'):
            m = re.match(r'\s*\d+\s+(\S+)\s+sprite=(-?\d+)\s+(\S+)', ln)
            if m and m.group(3) != '-':
                out[m.group(1)] = m.group(3)
    return out


# Ship hulls are not in the exe in any readable form (the built-in designs stay
# locked); the render-model dump in the research tree carries section geometry
# plus its own sprites, so a real hull can be drawn when that dump is present.
SHIP_HTML = os.path.join(RESEARCH, 'dump', 'ship_html')


def _hull_index(root=SHIP_HTML):
    """Squashed hull-dump name -> file.

    The dump is named for the *design* ("Space Station.html"), the mission names
    the *object* (`SpaceStation`), and nothing carries the mapping. Squashing
    both to lowercase letters and digits resolves the mechanical ones —
    SpaceStation, BattleStation, HestiaAlpha, MoiraX, OenoneDelta — which is
    every station and every ship variant EP9 can spawn. The rest (a platform
    with no object of its own, "Major Hong s Ship") have no object name to match
    and fall back to the object's sprite, as before.
    """
    out = {}
    if not os.path.isdir(root):
        return out
    for f in os.listdir(root):
        if f.endswith('.html') and f != 'index.html':
            out[re.sub(r'[^a-z0-9]', '', f[:-5].lower())] = os.path.join(root, f)
    return out


_HULLS = None


def _hull_path(name):
    """Where `name`'s hull dump is, or '' if it has none.

    Two steps: the dump named for the object exactly, then the squashed index.
    Both `ship()` and `hull_objects()` ask through here rather than each
    spelling the rule out, because they must agree — `hull_objects()` is what
    tells the editor to draw something as a ship, and `ship()` is what has to
    serve the hull when it does. Two transcriptions of the fallback is one
    chance for the editor to offer a hull thumbnail that then 404s.
    """
    global _HULLS
    if _HULLS is None:
        _HULLS = _hull_index()
    p = os.path.join(SHIP_HTML, name.replace(' ', '_') + '.html')
    if os.path.exists(p):
        return p
    p = _HULLS.get(re.sub(r'[^a-z0-9]', '', name.lower()), '')
    return p if p and os.path.exists(p) else ''


def hull_objects(names):
    """Which of `names` are ships rather than props.

    The distinction the editor needs up front. An object's `sprite_index` does
    not carry it: every stock hull's is `spr_Core`, the core section, because a
    ship draws its sections and never its own sprite — so the sprite manifest
    alone makes the Hestia, the Athena and a hundred others indistinguishable
    from each other and describes none of them. Having the hull is the test.
    """
    return sorted(n for n in names if _hull_path(n))


def ship(name):
    import json
    p = _hull_path(name)
    if not p:
        return None
    s = open(p, encoding='utf-8').read()
    def grab(const):
        m = re.search(r'const %s\s*=\s*(\{.*?\});\s*\n' % const, s, re.S)
        return json.loads(m.group(1)) if m else None
    sh, sp = grab('SHIP'), grab('SPRITES')
    if not sh or not sp:
        return None
    sh['sprites'] = sp
    return sh


def main():
    game = sys.argv[1] if len(sys.argv) > 1 else None
    lib = Library.from_game(game)
    print(f'{len(lib.sprites)} sprites from {lib.source}')
    if len(sys.argv) > 2:
        nm = sys.argv[2]
        s = lib.sprites[nm]
        png = lib.png(nm)
        print(f'{nm}: {s["w"]}x{s["h"]} origin=({s["ox"]},{s["oy"]}) frames={s["n"]} '
              f'mask={s["mask"]} png={len(png)}B')
    else:
        for nm in ['spr_Rock', 'spr_Nebula', 'spr_Spacestation', 'spr_Planet',
                   'spr_Meteor', 'spr_Marker', 'spr_MesHQ', 'spr_Section01']:
            if nm in lib.sprites:
                s = lib.sprites[nm]
                lib.png(nm)
                print(f'  {nm:20s} {s["w"]:>4}x{s["h"]:<4} origin=({s["ox"]},{s["oy"]})'
                      f' frames={s["n"]:<3} mask={s["mask"]}')


if __name__ == '__main__':
    main()
